torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+                        p._shard_bwd_hook[1].remove()  # type: ignore[attr-defined]
+                        delattr(p, "_shard_bwd_hook")
+
+        # Update root and nested FSDP's hooks and flags.
+        for m in self.modules():  # includes self
+            if isinstance(m, FullyShardedDataParallel):
+                _remove_shard_bwd_hook(m)
+                m._pre_backward_hook_has_run = False
+                if any(p.requires_grad for p in m.parameters()):
+                    # Check if the module has params and if any of them has
+                    # the `requires_grad` field set. If `requires_grad=False` for
+                    # all the params, the post_backward hook will not fire and the
+                    # state will remain in `TrainingState.BACKWARD_PRE`.
+                    if any([p.requires_grad for p in m.params]):
+                        m._assert_state(TrainingState.BACKWARD_POST)
+                    else:
+                        m._assert_state(TrainingState.BACKWARD_PRE)
+                else:
+                    # When `m` and its children has no params or has params but
+                    # none with `requires_grad==True`, there are two cases:
+                    # 1. output tensors are `requires_grad==True`. In this case,
+                    # pre-backward hook is still registered, so it is in BACKWARD_PRE state.
+                    # 2. output tensors are `requires_grad==False`. In this case,
+                    # pre-backward hook is not registered, so it is in IDLE state.
+                    m._assert_state([TrainingState.BACKWARD_PRE, TrainingState.IDLE])
+                m.training_state = TrainingState.IDLE
+
+                if m._is_root:
+                    # reset this flag for cases like "one forward pass + multiple backward passes"
+                    self._post_backward_callback_queued = False
+
+    @torch.no_grad()
+    def _rebuild_full_params(self) -> None:
+        """
+        Gather all shards of params.
+        """
+        def update_p_data(output_tensor: torch.Tensor) -> None:
+            """
+            Helper function to update p.data pointer.
+            Args:
+                output_tensor (torch.Tensor): this tensor contains the data we just gathered.
+            """
+            p.data = output_tensor
+            # Trim any padding and reshape to match original size.
+            p.data = p.data[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
+
+        with torch.cuda.stream(self._streams["all_gather"]):
+            for p in self.params:
+                # e.g., when world_size == 1
+                if not p._is_sharded:  # type: ignore[attr-defined]
+                    continue
+                else:
+                    # Move params in CPU to CUDA for the all-gather.
+                    p_data = p.data.to(p._full_param_padded.device, non_blocking=True)  # type: ignore[attr-defined]
```

Comment:
Is the data in pinned memory here?

Reply:
Yes, when we add cpu_offloading support, the cpu params and grads will be in pinned memory, let me remove this for now and will add it back when we add cpu_offloading